osu.Game/Screens/Edit/Compose/HitObjectUsageEventBuffer.cs

```diff
@@ -51,66 +51,23 @@ public HitObjectUsageEventBuffer([NotNull] Playfield playfield)
             playfield.HitObjectUsageFinished += onHitObjectUsageFinished;
         }
 
-        private readonly Dictionary<HitObject, EventType> pendingEvents = new Dictionary<HitObject, EventType>();
+        private readonly List<HitObject> usageFinishedHitObjects = new List<HitObject>();
 
-        private void onHitObjectUsageBegan(HitObject hitObject) => updateEvent(hitObject, EventType.Began);
-
-        private void onHitObjectUsageFinished(HitObject hitObject) => updateEvent(hitObject, EventType.Finished);
-
-        private void updateEvent(HitObject hitObject, EventType newEvent)
+        private void onHitObjectUsageBegan(HitObject hitObject)
         {
-            if (!pendingEvents.TryGetValue(hitObject, out EventType existingEvent))
-            {
-                pendingEvents[hitObject] = newEvent;
-                return;
-            }
-
-            switch (existingEvent, newEvent)
-            {
-                // This exists as a safeguard to ensure that the sequence: { Began -> Finished }, where { ... } indicates a sequence within a single frame, does not trigger any events.
-                // This is unlikely to occur in practice as it requires the usage to finish immediately after the HitObjectContainer updates hitobject lifetimes,
-                // however, an Editor action scheduled somewhere between the lifetime update and this buffer's own Update() could cause this.
-                case (EventType.Began, EventType.Finished):
-                    pendingEvents.Remove(hitObject);
-                    break;
-
-                // This exists as a safeguard to ensure that the sequence: Began -> { Finished -> Began -> Finished }, where { ... } indicates a sequence within a single frame,
-                // correctly leads into a final "finished" state rather than remaining in the intermediate "transferred" state.
-                // As above, this is unlikely to occur in practice.
-                case (EventType.Transferred, EventType.Finished):
-                    pendingEvents[hitObject] = EventType.Finished;
-                    break;
-
-                case (EventType.Finished, EventType.Began):
-                    pendingEvents[hitObject] = EventType.Transferred;
-                    break;
-
-                default:
-                    throw new ArgumentOutOfRangeException($"Unexpected event update ({existingEvent} => {newEvent}).");
-            }
+            if (usageFinishedHitObjects.Remove(hitObject))
+                HitObjectUsageTransferred?.Invoke(hitObject, playfield.AllHitObjects.Single(d => d.HitObject == hitObject));
```

Comment:
Not blocking this PR but I wonder `Playfield` should provide `HitObject` -> `DrawableHitObject` mapping public.

Reply:
I agree. This line in particular can probably benefit from optimisation as it's `O(n^2)`, but that also requires HOC changes + #12627.